packages/dates/src/format.ts

```diff
@@ -0,0 +1,155 @@
+  era?: 'long' | 'short' | 'narrow';
+  year?: 'numeric' | '2-digit';
+  month?: 'numeric' | '2-digit' | 'long' | 'short' | 'narrow';
+  day?: 'numeric' | '2-digit';
+  hour?: 'numeric' | '2-digit';
+  minute?: 'numeric' | '2-digit';
+  second?: 'numeric' | '2-digit';
+  timeZoneName?: 'long' | 'short';
+  formatMatcher?: 'basic' | 'best fit';
+  hour12?: boolean;
+  timeZone?: string;
+}
+
+type Token =
+  | 'YYYY'
+  | 'YY'
+  | 'MMMM'
+  | 'MMM'
+  | 'MM'
+  | 'M'
+  | 'DD'
+  | 'D'
+  | 'HH'
+  | 'H'
+  | 'hh'
+  | 'h'
+  | 'mm'
+  | 'm'
+  | 'ss'
+  | 's'
+  | 'A'
+  | 'a'
+  | 'dddd'
+  | 'ddd';
+
+type TokenDateTimeFormatOptionsMap = {
+  [key in Token]: {
+    type: Intl.DateTimeFormatPartTypes;
+    options: DateTimeFormatOptions;
+    extraParser?: (str: string) => string;
+  };
+};
+
+const TOKEN_MAP: TokenDateTimeFormatOptionsMap = {
+  YYYY: {type: 'year', options: {year: 'numeric'}},
+  YY: {type: 'year', options: {year: '2-digit'}},
+  MMMM: {type: 'month', options: {month: 'long'}},
+  MMM: {type: 'month', options: {month: 'short'}},
+  MM: {type: 'month', options: {month: '2-digit'}},
+  // eslint-disable-next-line id-length
+  M: {type: 'month', options: {month: 'numeric'}},
+  DD: {type: 'day', options: {day: '2-digit'}},
+  // eslint-disable-next-line id-length
+  D: {type: 'day', options: {day: 'numeric'}},
+  HH: {type: 'hour', options: {hour: '2-digit', hour12: false}},
+  // eslint-disable-next-line id-length
+  H: {
+    type: 'hour',
+    options: {hour: 'numeric', hour12: false},
+    // extraParser is needed when Node and Browsers don't respect the 2-digit
+    // and numeric options, which happens occasionally for time units
+    extraParser: numericFromTwoDigit,
+  },
+  hh: {type: 'hour', options: {hour: '2-digit', hour12: true}},
+  // eslint-disable-next-line id-length
+  h: {type: 'hour', options: {hour: 'numeric', hour12: true}},
+  mm: {
+    type: 'minute',
+    options: {minute: '2-digit'},
+    // see 'H' token parsing for more details about extraParser
+    extraParser: twoDigitFromNumeric,
+  },
+  // eslint-disable-next-line id-length
+  m: {type: 'minute', options: {minute: 'numeric'}},
+  ss: {
+    type: 'second',
+    options: {second: '2-digit'},
+    extraParser: twoDigitFromNumeric,
+  },
+  // eslint-disable-next-line id-length
+  s: {type: 'second', options: {second: 'numeric'}},
+  // eslint-disable-next-line id-length
+  A: {
+    type: 'dayPeriod',
+    options: {
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: true,
+    },
+  },
+  // eslint-disable-next-line id-length
+  a: {
+    type: 'dayPeriod',
+    options: {
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: true,
+    },
+    extraParser: lowercase,
+  },
+  dddd: {type: 'weekday', options: {weekday: 'long'}},
+  ddd: {type: 'weekday', options: {weekday: 'short'}},
+};
+
+function numericFromTwoDigit(time: string) {
```

Comment:
I would bundle these parsers into a `const PARSER_MAP = {` object above the TOKEN_MAP definition to signify that these are not just floating functions that are used elsewhere, but they are just parsers needed by the TOKEN_MAP. (Also despite hoisting, I personally find it easier to reason about code when functions that are needed are defined above, not below)

Reply:
Yeah, that's a good point. I myself had that thought when coming back to this PR after a day or two haha.

The bundling is also a good suggestion. Thanks!